metric_tank/defcache.go

```diff
@@ -0,0 +1,134 @@
+package main
+
+import (
+	"errors"
+	"github.com/grafana/grafana/pkg/log"
+	"github.com/raintank/raintank-metric/metricdef"
+	"github.com/raintank/raintank-metric/schema"
+	"sync"
+	"time"
+)
+
+// design notes:
+// MT pulls in all definitions when it starts up.
+// those "old" ones + whatever it sees as inputs from the metrics queue
+// is enough for it to always know the complete current state
+// nothing should update ES "behind its back", so we never need to pull
+// from ES other then at startup.
+// but other MT instances may update ES while we are down, so ES is a good
+// place to pull from, until the performance is demonstrably too slow.
+// there are some vectors here for race conditions but we can work those out
+// later, perhaps when tacking the multiple-intervals work
+
+type DefCache struct {
+	sync.RWMutex
+	defs map[string]*schema.MetricDefinition
+}
+
+func NewDefCache() *DefCache {
+	d := &DefCache{
+		defs: make(map[string]*schema.MetricDefinition),
+	}
+	go d.Backfill()
+	return d
+}
+
+// backfill definitions from ES
+// in theory, there is a race between defs from ES and from nsq
+// in practice, it doesn't matter: you're only supposed to query MT
+// after a while, after which the defs surely have stabilized.
+func (dc *DefCache) Backfill() {
+	total := 0
+	add := func(met []*schema.MetricDefinition) {
+		if len(met) > 0 {
+			total += len(met)
+			dc.Lock()
+			for _, def := range met {
+				dc.defs[def.Id] = def
+			}
+			dc.Unlock()
+		}
+	}
+	met, scroll_id, err := metricdef.GetMetrics("")
+	if err != nil {
+		log.Error(3, "Could not backfill from ES: %s", err)
+		return
+	}
+	add(met)
+	for scroll_id != "" {
+		met, scroll_id, err = metricdef.GetMetrics(scroll_id)
+		if err != nil {
+			log.Error(3, "Could not backfill from ES: %s", err)
+			return
+		}
+		add(met)
+	}
+	log.Debug("backfilled %d metric definitions", total)
+}
+
+func (dc *DefCache) Add(metric *schema.MetricData) {
+	id := metric.GetId()
+	dc.Lock()
+	mdef, ok := dc.defs[id]
+	dc.Unlock()
+	if ok {
+		if mdef.LastUpdate < time.Now().Unix()-600 {
```

Comment:
right. and now that i think about it. we should probably make sure only 1 tank service writes to ES, no need to multiply the load. maybe we can tie it to clusterstatus of primary. but then there could be a race condition of a metricdef update coming in during the brief transition where there is no primary. we can't use the metricpersist stream either cause it's semantics are just too different. so i'm thinking of 2 solutions:

A) a metricdefpersist stream
B) another clusterstatus property to toggle ES writing on/off. generally only 1 instance would have this enabled, but when we transition the duty to another machine, it would be ok to temporarily have 2 instances with this enabled (and make sure they're both processing their nsqd's properly).    we just need to make sure to have another istance up when the es-writer crashes, to be sure we didn't miss any metricdef, and when we promote-for-es-writing the new instance it can flush the updates to ES, but we'd still need to tell it something like "only flush updates as of <timestamp>" I think cause otherwise it can't know which defs in ES are up to date and which aren't, but A seems nicer for that. now that i think about it maybe we need both A and B.